tests/test_api.py

```diff
@@ -0,0 +1,374 @@
+    spec = ResolvingParser(str(here / 'data' / 'openapi-2.yaml'))
+    nb = AIONetbox(host='http://localhost', api_key='11', spec=spec, session=SessionMock())
+    nb.request = asynctest.CoroutineMock(return_value=ResponseMock())
+    op = NetboxApiOperation('users', 'users_list', nb.config['users']['users_list'], nb)
+
+    response_schema = spec.specification['paths']['/users']['get']['responses']['200']['schema']
+    op.parse_params = MagicMock(return_value=({}, {}, {'name': 'test'}))
+    op.build_url = MagicMock(return_value='http://test/users')
+
+    nbresp = await op.request(name='test', cf_hello='world', extra='ignore')
+
+    nb.request.assert_called_with(
+        method='get',
+        url='http://test/users',
+        query_params={'name': 'test', 'cf_hello': 'world'},
+        body={}
+    )
+
+    mnbro.from_response.assert_called_with(data={}, **response_schema)
+
+
+def test_NetboxApi():
+    spec = ResolvingParser(str(here / 'data' / 'openapi-2.yaml'))
+    nb = AIONetbox(host='http://localhost', api_key='11', spec=spec, session=SessionMock())
+    api = NetboxApi('users', nb.config.get('users'), nb)
+
+    assert repr(api) == 'Netbox.users'
+
+
+@patch('aionetbox.api.NetboxApiOperation')
+def test_NetboxApi_attr(mnbaop):
+    spec = ResolvingParser(str(here / 'data' / 'openapi-2.yaml'))
+    nb = AIONetbox(host='http://localhost', api_key='11', spec=spec, session=SessionMock())
+
+    api = NetboxApi('users', nb.config.get('users'), nb)
+
+    api.users_read
+    api.users_read
+
+    mnbaop.assert_called_once()
+    mnbaop.assert_called_with('users', 'users_read', nb.config['users']['users_read'], nb)
+
+    with pytest.raises(AttributeError):
+        api.fake_method
+
+
+@patch('aionetbox.api.ResolvingParser')
+def test_AIONetbox_from_openapi(mrp):
+    with patch.object(AIONetbox, 'parse_spec') as mps:
+        a = AIONetbox.from_openapi('http://testurl', 'ab12cd34')
+        mrp.assert_called_with('http://testurl/api/swagger.json')
+        mps.assert_called_with(mrp.return_value)
+
+
+def test_AIONetbox_from_spec():
+    a = AIONetbox.from_spec(str(here / 'data' / 'openapi-2.yaml'), 'ab12cd34')
+
+    assert a.host == 'https://api.example.com'
+
+
+def test_AIONetbox_parse_spec():
+    a = AIONetbox('http://localhost', '1122')
+    spec = ResolvingParser(str(here / 'data' / 'openapi-2.yaml'))
+
+    assert {} == a.parse_spec(None)
+    assert {} == a.parse_spec({})
+    assert '_orig' in a.parse_spec(spec)
+    assert 'users' in a.parse_spec(spec)
+    assert 'users_list' in a.parse_spec(spec).get('users')
+
+
+@pytest.mark.asyncio
+async def test_AIONetbox_request():
+    s = SessionMock()
+    a = AIONetbox('http://localhost', '1122', session=s)
+
+    with pytest.raises(ValueError):
+        await a.request('get', 'http://foobar', post_params={'hello': 'world'}, body={'wrong': True})
+
+    with pytest.raises(Exception):
+        await a.request('put', 'http://foobar', headers={'Content-Type': 'broken'})
```

Comment:
perhaps a matter of preference, but I'd separate this case out into three separate cases since it seems like its testing 2 different failure cases and 1 success case

Reply:
also, the check of `assert_called_with` isn't checked for the error cases here, which it could be (though I *think* the mock request wouldn't be called, in which case `assert_not_called` could be used)